pkg/kademlia/kademlia.go

```diff
@@ -0,0 +1,405 @@
+					k.waitNextMu.Unlock()
+
+					// TODO: somehow keep track of attempts and at some point forget about the peer
+					return false, false, nil // dont stop, continue to next peer
+				}
+
+				k.connectedPeers.Add(peer, po)
+
+				k.waitNextMu.Lock()
+				delete(k.waitNext, peer.String())
+				k.waitNextMu.Unlock()
+
+				k.depthMu.Lock()
+				k.depth = k.recalcDepth()
+				k.depthMu.Unlock()
+
+				k.logger.Debugf("connected to peer: %s old depth: %d new depth: %d", peer, currentDepth, k.NeighborhoodDepth())
+
+				// the bin could be saturated or not, so a decision cannot
+				// be made before checking the next peer, so we iterate to next
+				return false, false, nil
+			})
+
+			if err != nil {
+				if errors.Is(err, errMissingAddressBookEntry) {
+					po := uint8(swarm.Proximity(k.base.Bytes(), peerToRemove.Bytes()))
+					k.knownPeers.Remove(peerToRemove, po)
+				} else {
+					k.logger.Errorf("kademlia manage loop iterator: %v", err)
+				}
+			}
+		}
+	}
+}
+
+// binSaturated indicates whether a certain bin is saturated or not.
+// when a bin is not saturated it means we would like to proactively
+// initiate connections to other peers in the bin.
+func binSaturated(bin, depth uint8, peers *pslice.PSlice) bool {
+	// short circuit for bins which are >= depth
+	if bin >= depth {
+		return false
+	}
+
+	// lets assume for now that the minimum number of peers in a bin
+	// would be 2, under which we would always want to connect to new peers
+	// obviously this should be replaced with a better optimization
+	// the iterator is used here since when we check if a bin is saturated,
+	// the plain number of size of bin might not suffice (for example for squared
+	// gaps measurement)
+
+	size := 0
+	_ = peers.EachBin(func(_ swarm.Address, po uint8) (bool, bool, error) {
+		switch {
+		case po < bin:
+			return true, false, nil
+		case po > bin:
+			return false, true, nil
+		}
+
+		size++
+		return false, false, nil
+	})
+
+	return size >= 2
+}
+
+// recalcDepth returns the kademlia depth. Must be called under lock.
+func (k *Kad) recalcDepth() uint8 {
+	// handle edge case separately
+	if k.connectedPeers.Length() <= nnLowWatermark {
+		return 0
+	}
+	var (
+		peers                        = uint(0)
+		candidate                    = uint8(0)
+		shallowestEmpty, noEmptyBins = k.connectedPeers.ShallowestEmpty()
+	)
+
+	_ = k.connectedPeers.EachBin(func(_ swarm.Address, po uint8) (bool, bool, error) {
+		peers++
+		if peers >= nnLowWatermark {
+			candidate = po
+			return true, false, nil
+		}
+		return false, false, nil
+	})
+
+	if noEmptyBins || shallowestEmpty > candidate {
+		return candidate
+	}
+
+	return shallowestEmpty
+}
+
+// AddPeer adds a peer to the knownPeers list.
+// This does not guarantee that a connection will immediately
+// be made to the peer.
+func (k *Kad) AddPeer(ctx context.Context, addr swarm.Address) error {
+	if k.connectedPeers.Exists(addr) || k.knownPeers.Exists(addr) {
```

Comment:
well, this is certainly possible and should not be punishable, for example you might send me information about peers that i already know about from another peer (this is certainly possible, especially within nearest neighborhood) that informed me about them. certainly possible and should be fine

Reply:
yeah i agree..

Reply:
just loud thinking. May be we should increment a counter here when we get an `addPeer()` for a peer in known Peer List. This can help us to pick a highly connected node ([centrality](https://www.google.com/search?client=ubuntu&channel=fs&q=centrality&ie=utf-8&oe=utf-8)) when want to increase our saturation because someone other peer left in the bin.

Reply:
i think that picking a highly connected node could actually result in the reverse outcome, since you might end up creating more and more connection on that node, resulting in more load on that node (since it tries to serve more clients).

there is no trivial solution here, and this PR does not set out under the pretense to solve this problem. and it is indeed a relevant problem

Reply:
true. May be we can use this measure to avoid those node too. It will be good to add a //TODO: to take care of it in future.